HTTP/2 output processing make over

davidfowl opened the pull request:
This changes the HTTP/2's output processing to be use queues instead of locks (a channel to be precise). We're manually scheduling the Http2OutputProducer's intent to write instead of a write operation. This removes locks and allows the thread to do other useful work while processing in line for processing in a low allocation way.

Fixes #30235

~~Putting this up as a draft for now while I run performance tests and to get early feedback. ~~

## Scenaro: Unary invocation with 70 concurrent requests with one connection

### Current

| application                             |                                 |                          |
| --------------------------------------- | ------------------------------- | ------------------------ |
| CPU Usage (%)                           | 31                              |  ▃█▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▂   |
| Cores usage (%)                         | 867                             |  ▃█▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▂   |
| Working Set (MB)                        | 152                             | ███████████████████████  |
| Private Memory (MB)                     | 1,122                           | ▅▇█████████████████████  |
| Build Time (ms)                         | 14,507                          |                          |
| Start Time (ms)                         | 255                             |                          |
| Published Size (KB)                     | 91,501                          |                          |
| .NET Core SDK Version                   | 7.0.100-preview.3.22128.6       |                          |
| ASP.NET Core Version                    | 7.0.0-preview.4.22181.8+7399df2 |                          |
| .NET Runtime Version                    | 7.0.0-preview.4.22181.3+fe0f600 |                          |
| Max CPU Usage (%)                       | 31                              |   ▄█▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▆▂  |
| Max Working Set (MB)                    | 158                             | ████████████████████████ |
| Max GC Heap Size (MB)                   | 3                               | ▃▃▅█████████████████████ |
| Size of committed memory by the GC (MB) | 0                               |                          |
| Max Number of Gen 0 GCs / sec           | 0.00                            |                          |
| Max Number of Gen 1 GCs / sec           | 0.00                            |                          |
| Max Number of Gen 2 GCs / sec           | 0.00                            |                          |
| Max Time in GC (%)                      | 0.00                            |                          |
| Max Gen 0 Size (B)                      | 0                               |                          |
| Max Gen 1 Size (B)                      | 0                               |                          |
| Max Gen 2 Size (B)                      | 0                               |                          |
| Max LOH Size (B)                        | 0                               |                          |
| Max Allocation Rate (B/sec)             | 910,024                         |   █▃                     |
| Max GC Heap Fragmentation               | 0                               |                          |
| # of Assemblies Loaded                  | 110                             | ████████████████████████ |
| Max Exceptions (#/s)                    | 1                               |                       █  |
| Max Lock Contention (#/s)               | 16,862                          |   ▄█▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▁  |
| Max ThreadPool Threads Count            | 48                              |   ▅█████████████████████ |
| Max ThreadPool Queue Length             | 24                              |   █▄ ▂ ▁ ▃▁▂  ▇ ▁▁ ▃     |
| Max ThreadPool Items (#/s)              | 261,988                         |   ▂█▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▁  |
| Max Active Timers                       | 1                               | ████████████████████████ |
| IL Jitted (B)                           | 159,282                         | ▂▂▄▇▇▇▇▇▇▇██████████████ |
| Methods Jitted                          | 1,765                           | ▂▂▅▇▇▇▇▇▇▇▇▇▇▇██████████ |

| load                |           |                     |
| ------------------- | --------- | ------------------- |
| CPU Usage (%)       | 2         | ███████████████████ |
| Cores usage (%)     | 58        | █▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ |
| Working Set (MB)    | 46        | ▁▁▁▁▂▃▃▄▅▅▆▆▇██████ |
| Private Memory (MB) | 129       | ███████████████████ |
| Start Time (ms)     | 86        |                     |
| Requests            | 1,480,274 |                     |
| Bad responses       | 0         |                     |
| Socket errors       | 0         |                     |
| Mean latency (ms)   | 1         |                     |
| Max latency (ms)    | 18        |                     |
| Max RPS             | 98,685    |                     |

### PR

| application                             |                                 |                           |
| --------------------------------------- | ------------------------------- | ------------------------- |
| CPU Usage (%)                           | 33                              |   ▆███████████████████▁   |
| Cores usage (%)                         | 936                             |   ▅███████████████████▁   |
| Working Set (MB)                        | 153                             | ▇▇██████████████████████  |
| Private Memory (MB)                     | 959                             | ▆▆██████████████████████  |
| Build Time (ms)                         | 8,232                           |                           |
| Start Time (ms)                         | 269                             |                           |
| Published Size (KB)                     | 91,501                          |                           |
| .NET Core SDK Version                   | 7.0.100-preview.3.22128.6       |                           |
| ASP.NET Core Version                    | 7.0.0-preview.4.22181.8+7399df2 |                           |
| .NET Runtime Version                    | 7.0.0-preview.4.22181.3+fe0f600 |                           |
| Max CPU Usage (%)                       | 33                              |   ▅███████████████████▂   |
| Max Working Set (MB)                    | 159                             | ▇▇███████████████████████ |
| Max GC Heap Size (MB)                   | 3                               | ▃▃▅██████████████████████ |
| Size of committed memory by the GC (MB) | 0                               |                           |
| Max Number of Gen 0 GCs / sec           | 0.00                            |                           |
| Max Number of Gen 1 GCs / sec           | 0.00                            |                           |
| Max Number of Gen 2 GCs / sec           | 0.00                            |                           |
| Max Time in GC (%)                      | 0.00                            |                           |
| Max Gen 0 Size (B)                      | 0                               |                           |
| Max Gen 1 Size (B)                      | 0                               |                           |
| Max Gen 2 Size (B)                      | 0                               |                           |
| Max LOH Size (B)                        | 0                               |                           |
| Max Allocation Rate (B/sec)             | 1,206,304                       |   █▁                      |
| Max GC Heap Fragmentation               | 0                               |                           |
| # of Assemblies Loaded                  | 111                             | █████████████████████████ |
| Max Exceptions (#/s)                    | 1                               |                       █   |
| Max Lock Contention (#/s)               | 369                             |   ▄▆▇▆█▆▃▅▄▃▂▃▃▄▃▃▃▃▃▄▁   |
| Max ThreadPool Threads Count            | 28                              | ▁▁████████████████████▅▅▅ |
| Max ThreadPool Queue Length             | 2                               |   █   ▄    ▄       ▄▄     |
| Max ThreadPool Items (#/s)              | 417,645                         |   ▃▇██████████████████▂   |
| Max Active Timers                       | 1                               | █████████████████████████ |
| IL Jitted (B)                           | 234,850                         | ▂▂▆▇▇████████████████████ |
| Methods Jitted                          | 2,878                           | ▂▂▆▇▇████████████████████ |


| load                |           |                     |
| ------------------- | --------- | ------------------- |
| CPU Usage (%)       | 3         | ███████████████████ |
| Cores usage (%)     | 77        | ██▇████████████████ |
| Working Set (MB)    | 46        | ▁▁▁▁▃▅▇████████████ |
| Private Memory (MB) | 129       | ███████████████████ |
| Start Time (ms)     | 84        |                     |
| Requests            | 4,300,984 |                     |
| Bad responses       | 0         |                     |
| Socket errors       | 0         |                     |
| Mean latency (ms)   | 0         |                     |
| Max latency (ms)    | 1         |                     |
| Max RPS             | 286,732   |                     |

## Scenario: Unary invocation 1 stream and 1 connection

### Current

| load                |         |                     |
| ------------------- | ------- | ------------------- |
| CPU Usage (%)       | 1       | ███████████████████ |
| Cores usage (%)     | 34      | ███████████████████ |
| Working Set (MB)    | 13      | ▅▅▅▅▅▅▆▆▆▆▆▆▇▇▇▇███ |
| Private Memory (MB) | 129     | ███████████████████ |
| Start Time (ms)     | 89      |                     |
| Requests            | 145,094 |                     |
| Bad responses       | 0       |                     |
| Socket errors       | 0       |                     |
| Mean latency (ms)   | 0       |                     |
| Max latency (ms)    | 0       |                     |
| Max RPS             | 9,673   |                     |

### PR

| load                |         |                     |
| ------------------- | ------- | ------------------- |
| CPU Usage (%)       | 1       | ███████████████████ |
| Cores usage (%)     | 35      | ▇███▇████▇████▇████ |
| Working Set (MB)    | 13      | ▄▄▄▄▅▅▆▆▆▆▆▇▇▇▇▇▇██ |
| Private Memory (MB) | 129     | ███████████████████ |
| Start Time (ms)     | 88      |                     |
| Requests            | 147,366 |                     |
| Bad responses       | 0       |                     |
| Socket errors       | 0       |                     |
| Mean latency (ms)   | 0       |                     |
| Max latency (ms)    | 0       |                     |
| Max RPS             | 9,824   |                     |

davidfowl committed:
WIP

davidfowl committed:
Make more things work

davidfowl committed:
No more crashes!
- Fix double complete

davidfowl committed:
Added trailers

davidfowl committed:
Delete more code so it's obvious where changes are in the diff

davidfowl committed:
Complete the channel in Complete

davidfowl committed:
Move to flow control to the FrameWriter instead
- We only want to throttle data frames

davidfowl committed:
Use correct lock, add logs

davidfowl committed:
Abort flow control in a single place

davidfowl committed:
Remove dead code

davidfowl committed:
Undo chanes

davidfowl committed:
Remove call

davidfowl committed:
Made a few tests pass

davidfowl committed:
"Fixed" more tests

davidfowl committed:
Put back some timeouts

davidfowl committed:
Made more test fixes
- Allow scheduling the output for aborts and completions if we're waiting for window updates.

davidfowl committed:
Fix how we stop streams when flow control is being aborted

davidfowl committed:
Always clean up the pipe writer

davidfowl committed:
Fixed more tests

davidfowl committed:
Make inline scheduling work better
- Schedule before flushing so we can take advantage of the pipe scheduler

davidfowl committed:
Enqueue unflushed bytes before setting observation flag

davidfowl committed:
Moar fixes

davidfowl committed:
Cleaned up some tests

davidfowl committed:
Remove space

davidfowl committed:
More code cleanup

davidfowl committed:
Made the timeout tests work

davidfowl committed:
Tests pass!

davidfowl committed:
Move header writing to the processing loop
- Reverted the schedule before run pattern as it results in multiple dispatches for headers and body.
- Fixed tests to work with this pattern.

davidfowl committed:
Small clean up

davidfowl committed:
Merged the locks for header writing when possible

davidfowl committed:
More optimizations
- Write headers in the lock with the flush
- Added debug assert for scheduling output producers

davidfowl committed:
Remove a state machine from the write code path

davidfowl committed:
Rename firstWrite to writeHeaders

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +79,204 @@ internal class Http2FrameWriter
+        // This is null in tests sometimes
+        _window = connectionOutputFlowControl?.Available ?? 0;
+
+        _ = Task.Run(WriteToOutputPipe);
```
Should this task be stored on a field and awaited when closing a connection?

davidfowl replied:
Yes. I need to find a good place to do this. There's a risk things that haven't shutdown properly don't run if this isn't running. I had it on Complete before but that was too early since Abort calls Complete.

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +79,204 @@ internal class Http2FrameWriter
+
+    private async Task WriteToOutputPipe()
+    {
+        await foreach (var producer in _channel.Reader.ReadAllAsync())
```
Is the only way this loop exits the writer being completed? I think that's currently missing.

davidfowl replied:
https://github.com/dotnet/aspnetcore/pull/40925/files#r837053258

davidfowl replied:
Fixed

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +79,204 @@ internal class Http2FrameWriter
+                    buffer = buffer.Slice(0, actual);
+                }
+
+                var (hasModeData, scheduleAgain) = producer.Dequeue(buffer.Length, observed);
```
Extreme nit: scheduleAgain -> reschedule? 😆

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -125,7 +330,7 @@ public void Abort(ConnectionAbortedException error)
-    public ValueTask<FlushResult> FlushAsync(IHttpOutputAborter? outputAborter, CancellationToken cancellationToken)
+    private ValueTask<FlushResult> FlushAsync(Http2Stream stream, bool writeHeaders, IHttpOutputAborter? outputAborter, CancellationToken cancellationToken)
     {
         lock (_writeLock)
```
This is the lock that was causing contention problems. Can it be removed?

davidfowl replied:
Not fully. There will be less contention but window updates fire out of band from the normal processing so it's still required for that. Also writing 100 continue and probably some other obscure things.

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -411,102 +587,70 @@ void TrimAndWriteDataUnsynchronized(in ReadOnlySequence<byte> data, long dataLen
-                        // control updates and if the test checks for a timeout it can fail
-                        _timeoutControl.BytesWrittenToBuffer(_minResponseDataRate, _unflushedBytes);
-                    }
+                return new(default(FlushResult));
```
🤒 

Either `new ValueTask(...)` or `ValueTask.FromResult` 🙏

davidfowl replied:
I miss @pranavkm

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -125,7 +330,7 @@ public void Abort(ConnectionAbortedException error)
-    public ValueTask<FlushResult> FlushAsync(IHttpOutputAborter? outputAborter, CancellationToken cancellationToken)
+    private ValueTask<FlushResult> FlushAsync(Http2Stream stream, bool writeHeaders, IHttpOutputAborter? outputAborter, CancellationToken cancellationToken)
     {
         lock (_writeLock)
```
This is the lock that was causing contention problems. Can it be removed?

davidfowl committed:
Make the channel unboundeded
- We allow more than the max streams by a little bit. This won't be much more than limit but we don't need to risk dropping streams as a result.

davidfowl committed:
Shutdown the write queue after the connection loop ends

davidfowl committed:
Make James happy

davidfowl committed:
Rename schedule again to reschedule

adityamandaleeka commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +80,204 @@ internal class Http2FrameWriter
+                    buffer = buffer.Slice(0, actual);
+                }
+
+                var (hasModeData, reschedule) = producer.Dequeue(buffer.Length, observed);
```
```suggestion
                var (hasMoreData, reschedule) = producer.Dequeue(buffer.Length, observed);
```

adityamandaleeka commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs:
```diff
@@ -66,20 +70,103 @@ public Http2OutputProducer(Http2Stream stream, Http2StreamContext context, Strea
+    {
+        lock (_dataWriterLock)
+        {
+            var wasDepleted = _window <= 0;
```
Can `_window` ever go below 0?

davidfowl replied:
Yes, guess how I found out... The tests!

davidfowl committed:
Rename cancelled to use one l

davidfowl committed:
Spelling

davidfowl committed:
Move trailer writing into a single method

davidfowl committed:
Remove extra lines...

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -173,15 +173,8 @@ await foreach (var producer in _channel.Reader.ReadAllAsync())
-                    {
-                        flushResult = await WriteResponseTrailersAsync(stream, producer.WriteHeaders, stream.ResponseTrailers);
-                    }
+                    // It is faster to write data and trailers together. Locking once reduces lock contention.
+                    flushResult = await WriteDataAndTrailersAsync(stream, buffer, producer.WriteHeaders, stream.ResponseTrailers);
```
I combined these to reduce grpc requests fighting over the lock. Theoretically, they don't need to be merged anymore if only one thread is taking the write lock (and the occasional window updates).

I don't think anything needs to be done now, should wait for perf data first, but maybe they should be separate methods.

davidfowl replied:
I don't think so. It's actually nicer have them all combined into a single method.

davidfowl committed:
Remove task chain allocation on shutdown
- We were waiting on the response completion task in write suffix, this caused us to allocate an set of async state machines while unwinding the request processing. This changes the flow to avoid that await completely and leave the clean up of the stream to either the request processing loop or framewriter queue.

davidfowl committed:
Removed the use of the output flow control objects from the source

davidfowl committed:
Delete output flow control types

davidfowl committed:
Make ConsumeWindow private

davidfowl committed:
Remove if from queue processing

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +77,189 @@ internal class Http2FrameWriter
+                // We don't need to check the result because it's either
+                // - true because we have a result
+                // - false because we're flushing headers
+                var hasResult = reader.TryRead(out var readResult);
```
nit

Why not `reader.TryRead(out var readResult);` with no hasResult variable? Or is it a warning to not use the result?

Alternatively, `_ = reader.TryRead(out var readResult);` to discard.

davidfowl replied:
oops. Mistake.

davidfowl committed:
Oops

davidfowl committed:
Fixed logs

JamesNK committed:
Fix merge

JamesNK commented:
Fixed merge conflicts.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+                {
+                    if (buffer.Length != 0)
+                    {
+                        _log.LogCritical("{StreamId} observed an unexpected state where the streams output ended with data still remaining in the pipe.", stream.StreamId);
```
If we're not going to bother with `DecrementActiveClientStreamCount`, we should probably also abort the connection. I get this should never happen, but still.

davidfowl replied:
As long as I don't have to fix a boatload of tests. I preserved the behavior as best I could.

JamesNK replied:
Before this was a stream-level problem. An error was thrown and the stream output loop would exit: https://github.com/dotnet/aspnetcore/blob/e9b2adb2a21cc2b0ae5b8f7e93aa4599d6f847b0/src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs#L438

If we preserve that behavior then the stream/connection needs to handle the broken stream. Not decremeanting the active stream count will hang a connection if this happens often enough

Alternatively, if this is something that shouldn't ever happen then aborting the connection is a simpler Just In Case.

davidfowl replied:
I'm confused. I'm 96% sure the code is as it was before. What's different about this? The old code used to complete the pipe reader and didn't abort anything. This is new behavior...

JamesNK replied:
This new loop operates at a connection level instead of stream level.

My question is, is the stream pipe reader still completed? Is the active stream count being decremented somewhere? (TBH I'm not sure where the active stream count would be decremented in this situation in old code)

At the very least I think a comment here that explains why it's ok that after logging the critical error that `DecrementActiveClientStreamCount` isn't called and the connection's not aborted.

davidfowl replied:
> My question is, is the stream pipe reader still completed?

No.

>  Is the active stream count being decremented somewhere?

When the stream completes by other means (like the request ending).

> At the very least I think a comment here that explains why it's ok that after logging the critical error that DecrementActiveClientStreamCount isn't called and the connection's not aborted.

I'm all for fixing this but I'm not planning to introduce new behavior in this PR.

davidfowl replied:
I will make sure it's the same logic that exists today.

davidfowl replied:
The pipe is cleaned up as expected. The behavior is equivalent. We can file an issue about making it abort the connection more promptly if need be.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs:
```diff
@@ -57,6 +57,10 @@ public ConcurrentPipeWriter(PipeWriter innerPipeWriter, MemoryPool<byte> pool, o
 
+    public override bool CanGetUnflushedBytes => true;
+
+    public override long UnflushedBytes => _currentFlushTcs is null && _head is null ? _innerPipeWriter.UnflushedBytes : _bytesBuffered;
```
Shouldn't this be?

```suggestion
    public override long UnflushedBytes => _currentFlushTcs is null && _head is null ? _innerPipeWriter.UnflushedBytes : _bytesBuffered + _innerPipeWriter.UnflushedBytes;
```

Also, I don't think `ConcurrentPipeWriter` is guaranteed to have an `_innerPipeWriter` where `CanGetUnflushedBytes` is true. For `Http1Connection,` it comes from `ConnectionContext.Transport`.

JamesNK replied:
If it should be this, and the current code passes all tests, then a test should be added to verify correctness here.

davidfowl replied:
Removed it

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs:
```diff
@@ -41,47 +34,126 @@ internal class Http2OutputProducer : IHttpOutputProducer, IHttpOutputAborter, IV
+    // Adds more bytes to the stream's window
+    // Returns a bool that represents whether we should schedule this producer to write
+    // the remaining bytes.
+    private bool UpdateWindow(long bytes)
```
```suggestion
    private bool UpdateStreamWindow(long bytes)
```

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+
+                        lock (_windowUpdateLock)
+                        {
+                            _waitingForMoreWindow.Enqueue(producer);
```
If a stream used all of the available stream window and connection window at once, the stream could get enqueued at least twice into `_waitingForMoreWindow` without ever being dequeued if it received a stream window update before a connection window update. This can't be repeated because of the `wasDepleted` check in `UpdateWindow()`, but I think it would be safest to check that `producer._waitingForWindowUpdates` isn't already set.

davidfowl replied:
Handled by schedule

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
         _hpackEncoder = new DynamicHPackEncoder(serviceContext.ServerOptions.AllowResponseHeaderCompression);
+
+        // In practice, this is bounded by the number of concurrent streams allowed + whatever overflow we allow
+        _channel = Channel.CreateUnbounded<Http2OutputProducer>(new UnboundedChannelOptions()
```
I'd rather debug assert something than bound this channel. Or I'd agree would could bound it to 2x + buffer or something like this.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -44,11 +45,16 @@ internal class Http2FrameWriter
 
+    private readonly object _windowUpdateLock = new();
+    private long _window;
+    private readonly Queue<Http2OutputProducer> _waitingForMoreWindow = new();
```
```suggestion
    private readonly Queue<Http2OutputProducer> _waitingForMoreConnectionWindow = new();
```

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -352,20 +486,29 @@ static async ValueTask<FlushResult> WriteDataAndTrailersAsyncCore(Http2FrameWrit
+
+        WriteHeaderUnsynchronized();
+
+        data.CopyTo(_outputWriter);
```
Why the inconsistency between `CopyTo` in some places an foreaching over the data in others?

davidfowl replied:
This is unchanged codez.

JamesNK replied:
Change it to CopyTo everywhere. It has an optimized path for single segment ROS.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs:
```diff
@@ -41,47 +34,126 @@ internal class Http2OutputProducer : IHttpOutputProducer, IHttpOutputAborter, IV
+    // Removes consumed bytes from the queue.
+    // Returns a bool that represents whether we should schedule this producer to write
+    // the remaining bytes.
+    internal (bool, bool) Dequeue(long bytes, State state)
```
```suggestion
    internal (bool HasMoreData, bool HasUnobservedState) ObserveDataAndState(long observedBytes, State observedState)
```

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+                    _timeoutControl.StopTimingWrite();
+                }
+
+                producer.WriteHeaders = false;
```
Is there any chance that the first time a write gets enqueued `WriteHeaders` is false but then becomes true later? I don't think so, but this unsynchronized reading and setting of the field has me nervous.

davidfowl replied:
It isn't but I don't have a problem changing this state under a lock.

davidfowl replied:
Fixed

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+            }
+            catch (Exception ex)
+            {
+                _log.LogCritical(ex, "The event loop in connection {ConnectionId} failed unexpectedly", _connectionId);
```
We should also abort the connection in this case.

davidfowl replied:
Is that what happens today?

JamesNK replied:
There isn't a connection-level loop like this today.

The closest equivalent is the stream-level loop in the output producer: https://github.com/dotnet/aspnetcore/blob/e9b2adb2a21cc2b0ae5b8f7e93aa4599d6f847b0/src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs#L456

It doesn't abort, but various pipes and completion sources are completed to indicate it's no longer running.

davidfowl replied:
Unfortunately it doesn't do what we think it does.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -725,27 +830,56 @@ private ValueTask<FlushResult> TimeFlushUnsynchronizedAsync()
     }
 
-    public bool TryUpdateStreamWindow(StreamOutputFlowControl flowControl, int bytes)
+    private void AbortFlowControl()
```
```suggestion
    private void AbortConnectionFlowControl()
```

halter73 reviewed (commented):
I'd like to be clearer about whether we're referring to the connection or stream window.

halter73 committed:
Don't reschedule for nothing

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -172,48 +372,42 @@ public ValueTask<FlushResult> Write100ContinueAsync(int streamId)
+            _headersEnumerator.Initialize(headers);
+            _outgoingFrame.PrepareHeaders(headerFrameFlags, streamId);
+            var buffer = _headerEncodingBuffer.AsSpan();
+            var done = HPackHeaderWriter.BeginEncodeHeaders(statusCode, _hpackEncoder, _headersEnumerator, buffer, out var payloadLength);
```
I believe that as long as only one HPACK header writer is operating at once (dynamic table is stateful) then it can happen outside of the write lock.

davidfowl replied:
Sure, this can be a follow up. My priority with this PR is to, avoid breaking anything and improve the performance.

davidfowl replied:
Will file an issue for this.

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -78,21 +79,33 @@ internal class Http2FrameWriter
+        if (!_channel.Writer.TryWrite(producer))
+        {
+            // It should not be possible to exceed the bound of the channel.
+            var ex = new ConnectionAbortedException("Http2FrameWriter._channel exceeded bounds.");
```
I'm not a fan of including names of fields in error messages that are logged.

What about something like: `Connection exceeded the number of allowed output operations.`

halter73 replied:
```suggestion
            var ex = new ConnectionAbortedException("HTTP/2 connection exceeded the output operations maximum queue size.");
```

I'm fine with this. My original proposal had `Environment.FailFast("Http2FrameWriter._channel exceeded bounds.")`, so this was a compromise. Hopefully no one ever sees this.

davidfowl replied:
No fail fast in the server side code please.

halter73 replied:
Understood. `Environment.FailFast` would make for an easy DoS if it's actually hittable. I just want to make sure that this channel is **really** bounded and not something that can grow without limit in some edge case, because the OOM killer can also DoS you server if this is in fact unbounded.

davidfowl replied:
I don't mind bounding the channel but dropping the stream doesn't need to crash the server process.

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+            }
+        }
+
+        _log.LogDebug("The connection processing loop for {ConnectionId} ended gracefully", _connectionId);
```
This is logged for every completed connection. Should use defined log via source gen for perf.

davidfowl replied:
Yep, I need to clean that up.

davidfowl commented:
@halter73 your last commit broke quite a few new tests.

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+                    observed |= Http2OutputProducer.State.Completed;
+                }
+
+                // Check the stream window
+                var (actual, remainingStream) = producer.ConsumeWindow(buffer.Length);
+
+                // Now check the connection window
+                (actual, var remainingConnection) = ConsumeWindow(actual);
```
I'm confused by this. What is the issue? I'm waiting more window updates and the application completes?

davidfowl replied:
Resolved, we check and only write 0 length data frames if we're at the end of stream

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs:
```diff
@@ -157,19 +249,52 @@ public ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken)
+        lock (_dataWriterLock)
+        {
+            // Lock here
+            _waitingForWindowUpdates = false;
```
Do we do this today?

halter73 replied:
I think `_isScheduled` is sufficient. I doubt the other case where we don't need to schedule because of lack of window are as common. If they are, we can optimize later.

halter73 committed:
Fix typo

davidfowl committed:
Apply PR feedback

davidfowl committed:
Removed unneeded state

davidfowl committed:
More PR suggestions

davidfowl committed:
Remove the FlushHeaders bool and use the unobserved state

davidfowl committed:
Small nits

davidfowl committed:
Small tweaks

davidfowl committed:
More naming changs

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -72,6 +77,197 @@ internal class Http2FrameWriter
+                    {
+                        if (buffer.Length != 0)
+                        {
+                            _log.LogCritical("{StreamId} observed an unexpected state where the streams output ended with data still remaining in the pipe.", stream.StreamId);
```
nit
```suggestion
                            _log.LogCritical("Stream {StreamId} observed an unexpected state where the stream's output ended with data still remaining in the pipe.", stream.StreamId);
```

davidfowl committed:
More PR feedback
- Added logging entries to source gen
- Moved UpdateStreamWindow to local function

davidfowl committed:
Removed usings...

davidfowl committed:
Aborts and writes are no longer synchronized
- Writes have not been written to the connection on complete async so aborting them can cancel pending data can be cancelled even if called after complete.

davidfowl committed:
Preserve existing behavior
- If we complete the stream before calling WriteRst then wait until the response completes before writing the RST

davidfowl committed:
Removed unused state

davidfowl committed:
Make the test await the process request loop

davidfowl commented:
PS: The numbers were obtained by using a 4K pause threshold for the output pipe, that isn't currently in this PR.

davidfowl committed:
Removed extra flush headers state.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs:
```diff
@@ -380,96 +509,60 @@ public void Stop()
+            {
+                // If we have an error code to write, write it now that we're done with the response.
+                // Always send the reset even if the response body is completed. The request body may not have completed yet.
+                _ = _frameWriter.WriteRstStreamAsync(StreamId, error).Preserve();
```
Can we make CompleteResponse() Task-returning and await it in WriteToOutputPipe?

davidfowl replied:
Done

halter73 commented on src/Servers/Kestrel/test/Interop.FunctionalTests/HttpClientHttp2InteropTests.cs:
```diff
@@ -637,6 +637,8 @@ public async Task ServerReset_AfterHeaders_ClientBodyThrows(string scheme)
     [MemberData(nameof(SupportedSchemes))]
     public async Task ServerReset_AfterEndStream_NoError(string scheme)
```
I would normally say I don't care about there not being an error if Reset()/Abort() is called before middleware completes, but CompleteAsync() does imply a certain finality. I'm glad that the behavior is preserved, but it's pretty edge-casey behavior, so it wouldn't have been the end of the world.

JamesNK replied:
FYI gRPC relies on this for deadline functionality. CompleteAsync is used to end the request, which in gRPC's case means to flush trailers with a status trailer, then Reset is used to send a RST_STREAM of NO_ERROR to the client.

davidfowl replied:
bizzare but good to know I guess...

davidfowl committed:
PR feedback

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.cs:
```diff
@@ -298,6 +298,38 @@ public void Http2MaxConcurrentStreamsReached(string connectionId)
+        Http2QueueOperationsExceeded(_http2Logger, connectionId, ex);
+    }
+
+    [LoggerMessage(61, LogLevel.Critical, @"Stream {StreamId} observed an unexpected state where the streams output ended with data still remaining in the pipe.", EventName = "Http2UnexpectedDataRemaining")]
```
Done

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -725,27 +830,56 @@ private ValueTask<FlushResult> TimeFlushUnsynchronizedAsync()
+                if (!producer.StreamCompleted)
+                {
+                    // We're no longer waiting for the update
+                    producer.MarkWaitingForWindowUpdates(false);
```
OK I spent some time with this today looking at the most minimal changes to basically replicate what OutputFlowControl did. Let me know if my assumptions are wrong in any of this.

The old code also scheduled all waiting streams as long as while there was still flow control window available. I don't see any special logic to deal with fairness outside of that. 

https://github.com/dotnet/aspnetcore/blob/1d922b9dbdaeafe7468094568f934cbcc8c8dc73/src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/OutputFlowControl.cs#L49-L62

So when CompleteCurrent is called, it queues the continuation to the threadpool and keeps doing so while there's flow window available. I'm missing how this avoids the thundering herd problem.

davidfowl replied:
Actually, looking at it more closely, since TryUpdateWindow for the connection level flow control is called within the write lock, I'm not sure how the available window can be decremented (via call to OutputFlowControl.Advance) since its also only allowed to be called inside of the write lock. It seems like all streams would need to be dequeued before any new stream can proceed since that happens while the lock is held.

I'm not sure `_flow.Available > 0` is ever false.

davidfowl replied:
I added a commit to do something minimal that I think solves the problem in the worst case. Please review and let me know if I missed something about the existing code.

davidfowl committed:
Improve the fairness of connection window updates
- Today its possible for a single stream to keep getting rescheduled and draining the connection window. This change makes it so that streams that have data to be written and haven't written anything are preferred over ones that have written something. It does this by changing the head of the queue before dequeuing everything to start at the first index of the stream that hasn't written anything.

davidfowl committed:
Fixed test

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/QueueWithMovableHead.cs:
```diff
@@ -0,0 +1,137 @@
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2;
+
+// This is a normal queue with a twist, we have the ability to set the head to a different start index
+internal class QueueWithMovableHead<T>
```
```suggestion
internal sealed class QueueWithMovableHead<T>
```

JamesNK commented on src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2StreamTests.cs:
```diff
@@ -3103,15 +3103,11 @@ public async Task ResponseWithHeadersTooLarge_AbortsConnection()
             context.Response.Headers["too_long"] = new string('a', (int)Http2PeerSettings.DefaultMaxFrameSize);
-            var ex = await Assert.ThrowsAsync<InvalidOperationException>(() => context.Response.WriteAsync("Hello World")).DefaultTimeout();
-            appFinished.TrySetResult(ex.InnerException.Message);
+            await context.Response.WriteAsync("Hello World");
```
This is a small behavior change. Is the new behavior intentional?

davidfowl replied:
Yes. There's no way to get a synchronous exception when writes are queued like this. I spoke to @Tratcher about it and he agreed this is fine.

halter73 replied:
We could preemptively check if any outgoing header values are longer than the current value of `Http2FrameWriter._maxFrameSize`. It makes this error a little clearer to the developer if it gets thrown directly to their app vs just showing up in the logs.

I don't think it's worth it though because we'd have to iterate over the response headers and additional time every request just for this validation.

davidfowl commented:
I'm attempting to make the pipe buffer change and fix one more round of tests... Wish me luck...

davidfowl committed:
Increased pause threshold to 4K
- Reworked logic in frameworker to use the state enum as well as the read result.
- Fixed races around how work can be scheduled now that we're not running inline with a small pause threshold.

davidfowl committed:
Simplify the queue fairness
- Store the last window consumer in as a field and make sure it gets queued last on the next window update.

davidfowl commented:
I removed the `QueueWithMovableHead ` and implemented the suggestion.

davidfowl committed:
Fixed 2 remaining issues
- Over scheduling races (cancel during a dequeue that will complete the pipe)
- Handle cancellation when there are trailers

davidfowl committed:
Added last window consumer to the queue when aborting

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs:
```diff
@@ -4,66 +4,57 @@
-    ValueTaskSourceStatus IValueTaskSource<FlushResult>.GetStatus(short token) => _responseCompleteTaskSource.GetStatus(token);
-    void IValueTaskSource<FlushResult>.OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags) => _responseCompleteTaskSource.OnCompleted(continuation, state, token, flags);
-    FlushResult IValueTaskSource<FlushResult>.GetResult(short token) => _responseCompleteTaskSource.GetResult(token);
+    // For changes scheduling changes that don't affect the number of bytes written to the pipe, we need another state
```
```suggestion
    // For scheduling changes that don't affect the number of bytes written to the pipe, we need another state
```

davidfowl committed:
Stop for RST frames as well.

davidfowl committed:
Small tweaks to state management
- Added current state to make sure state is terminal. We base decisions on the current state and use the unobserved state to determine when to reschedule.
- Remove CancelPendingRead since there's never a pending read and we're scheduling the pipe reads manually.
- Rename the Canceled state to Aborted

davidfowl committed:
This is the last commit that fixes all of the bugs (I hope...)

davidfowl committed:
Make things "simpler"

davidfowl committed:
We're not boxing today

davidfowl commented:
We're good @halter73 !

JamesNK commented:
Nice!

One request: run the gRPC benchmark with 70 streams and 28 connections.

I’d like to see how the change performs when the server is at capacity

davidfowl commented:
@JamesNK do you have a command for me to run? 😄

JamesNK commented:
It’s the same one but add “—variable connections=28”

davidfowl commented:
## 28 connections, 70 streams

### Before 

| application                             |                                 |                           |
| --------------------------------------- | ------------------------------- | ------------------------- |
| CPU Usage (%)                           | 83                              |   █▆▆▆▆▅▅▅▅▆▅▅▅▆▆▆▅▅▅▄    |
| Cores usage (%)                         | 2,313                           |   █▆▆▆▆▅▅▆▅▆▅▆▅▆▆▆▅▅▅▄    |
| Working Set (MB)                        | 73                              | ▅▆██████████████████████  |
| Private Memory (MB)                     | 104                             | ▆▇██████████████████████  |
| Build Time (ms)                         | 4,115                           |                           |
| Start Time (ms)                         | 402                             |                           |
| Published Size (KB)                     | 91,316                          |                           |
| .NET Core SDK Version                   | 7.0.100-preview.3.22128.6       |                           |
| ASP.NET Core Version                    | 7.0.0-preview.4.22215.9+3e8db65 |                           |
| .NET Runtime Version                    | 7.0.0-preview.4.22215.4+17c2cb2 |                           |
| Max CPU Usage (%)                       | 84                              |    █▆▆▆▆▅▅▆▅▆▅▅▅▆▆▆▅▅▆▄   |
| Max Working Set (MB)                    | 76                              | ▅▅▇██████████████████████ |
| Max GC Heap Size (MB)                   | 15                              | ▁▁███████████████████████ |
| Size of committed memory by the GC (MB) | 0                               |                           |
| Max Number of Gen 0 GCs / sec           | 0.00                            |                           |
| Max Number of Gen 1 GCs / sec           | 0.00                            |                           |
| Max Number of Gen 2 GCs / sec           | 0.00                            |                           |
| Max Time in GC (%)                      | 0.00                            |                           |
| Max Gen 0 Size (B)                      | 0                               |                           |
| Max Gen 1 Size (B)                      | 0                               |                           |
| Max Gen 2 Size (B)                      | 0                               |                           |
| Max LOH Size (B)                        | 0                               |                           |
| Max Allocation Rate (B/sec)             | 12,821,176                      |   █                       |
| Max GC Heap Fragmentation               | 0                               |                           |
| # of Assemblies Loaded                  | 112                             | █████████████████████████ |
| Max Exceptions (#/s)                    | 6                               |                       █   |
| Max Lock Contention (#/s)               | 123,493                         |    ▃▇▇▇█▇▇█▇▇▇█▇▇▇▇▇▆▆▆   |
| Max ThreadPool Threads Count            | 48                              |   ▅██████████████████████ |
| Max ThreadPool Queue Length             | 1,078                           |   █    ▃ ▁ ▁ ▁▁ ▂▁   ▁    |
| Max ThreadPool Items (#/s)              | 912,377                         |    ▆████▇▇▇▇▇▇▇▇▇▇▇▇▇▇▆   |
| Max Active Timers                       | 0                               |                           |
| IL Jitted (B)                           | 152,660                         | ▂▂▃▅▇████████████████████ |
| Methods Jitted                          | 1,777                           | ▂▂▃▅▇████████████████████ |


| load                |            |                     |
| ------------------- | ---------- | ------------------- |
| CPU Usage (%)       | 4          | ▆████▆████████▆█▆██ |
| Cores usage (%)     | 100        | ███████████████████ |
| Working Set (MB)    | 48         | ▂▂▂▂▆██████████████ |
| Private Memory (MB) | 131        | ███████████████████ |
| Start Time (ms)     | 82         |                     |
| Requests            | 10,365,490 |                     |
| Bad responses       | 0          |                     |
| Socket errors       | 0          |                     |
| Mean latency (ms)   | 2          |                     |
| Max latency (ms)    | 12         |                     |
| Max RPS             | 691,033    |                     |


### After

| application                             |                                 |                           |
| --------------------------------------- | ------------------------------- | ------------------------- |
| CPU Usage (%)                           | 96                              |  ▃█▅▅▅▅▄▄▄▅▅▄▅▄▅▅▅▅▅▄▁    |
| Cores usage (%)                         | 2,686                           |  ▃█▅▅▅▅▄▄▄▅▅▄▅▄▅▅▅▅▅▄▁    |
| Working Set (MB)                        | 79                              | ▅███████████████████████  |
| Private Memory (MB)                     | 110                             | ▆███████████████████████  |
| Build Time (ms)                         | 4,143                           |                           |
| Start Time (ms)                         | 420                             |                           |
| Published Size (KB)                     | 91,316                          |                           |
| .NET Core SDK Version                   | 7.0.100-preview.3.22128.6       |                           |
| ASP.NET Core Version                    | 7.0.0-preview.4.22215.9+3e8db65 |                           |
| .NET Runtime Version                    | 7.0.0-preview.4.22215.4+17c2cb2 |                           |
| Max CPU Usage (%)                       | 98                              |   ▄█▅▅▅▅▄▄▄▄▄▄▅▄▄▅▅▄▅▄▂   |
| Max Working Set (MB)                    | 82                              | ▅▅███████████████████████ |
| Max GC Heap Size (MB)                   | 22                              | ▁▁███████████████████████ |
| Size of committed memory by the GC (MB) | 0                               |                           |
| Max Number of Gen 0 GCs / sec           | 0.00                            |                           |
| Max Number of Gen 1 GCs / sec           | 0.00                            |                           |
| Max Number of Gen 2 GCs / sec           | 0.00                            |                           |
| Max Time in GC (%)                      | 0.00                            |                           |
| Max Gen 0 Size (B)                      | 0                               |                           |
| Max Gen 1 Size (B)                      | 0                               |                           |
| Max Gen 2 Size (B)                      | 0                               |                           |
| Max LOH Size (B)                        | 0                               |                           |
| Max Allocation Rate (B/sec)             | 20,203,608                      |   █                       |
| Max GC Heap Fragmentation               | 0                               |                           |
| # of Assemblies Loaded                  | 113                             | █████████████████████████ |
| Max Exceptions (#/s)                    | 0                               |                           |
| Max Lock Contention (#/s)               | 968                             |   ▂█▇██▇▇▇▇▇▇▆▇▆▇▆▆▇▇▆▃   |
| Max ThreadPool Threads Count            | 48                              |   ▅██████████████████████ |
| Max ThreadPool Queue Length             | 399                             |   █                       |
| Max ThreadPool Items (#/s)              | 950,535                         |   ▂▆████▇▇▇▇▇▇▇▇▇▇▇▇▇▇▃   |
| Max Active Timers                       | 0                               |                           |
| IL Jitted (B)                           | 232,817                         | ▂▂▄▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇███ |
| Methods Jitted                          | 2,909                           | ▂▂▄▇▇████████████████████ |


| load                |            |                     |
| ------------------- | ---------- | ------------------- |
| CPU Usage (%)       | 4          | ███████████▆███████ |
| Cores usage (%)     | 101        | ███████████████████ |
| Working Set (MB)    | 48         | ▂▂▂▂▆██████████████ |
| Private Memory (MB) | 131        | ███████████████████ |
| Start Time (ms)     | 86         |                     |
| Requests            | 11,517,940 |                     |
| Bad responses       | 0          |                     |
| Socket errors       | 0          |                     |
| Mean latency (ms)   | 1          |                     |
| Max latency (ms)    | 12         |                     |
| Max RPS             | 767,863    |                     |

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:
```diff
@@ -74,6 +71,7 @@ internal partial class Http2Connection : IHttp2StreamLifetimeHandler, IHttpStrea
     internal readonly Http2KeepAlive? _keepAlive;
     internal readonly Dictionary<int, Http2Stream> _streams = new Dictionary<int, Http2Stream>();
     internal PooledStreamStack<Http2Stream> StreamPool;
+    internal Action<Http2Stream>? _onStreamCompleted;
```
Is this just for testing?

I don't think it needs to be fixed now, but I'd like to see this reworked to utilize `IHttp2StreamLifetimeHandler`. Tests should support customizing the handler so it can be used to capture this event. I did this in the HTTP/3 tests and it works well.

JamesNK replied:
Maybe create a follow-up issue.

JamesNK reviewed (approved):
🚢 🇮🇹

davidfowl commented:
```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/grpc.benchmarks.yml --scenario grpcaspnetcoreserver-h2loadclient --profile aspnet-citrine-win --variable streams=70 --variable connections=1 --application.framework net7.0 --application.sdkversion 7.0.100-preview.3.22128.6 --application.options.collectCounters true --chart
```
